File: uiweb/src/debug/MiniCDebugTextFormatter.ts
import type { JavaMirrorFile } from "../translation/javaMirror";
import type {
  UiDebugBreakpointDto,
  UiDebugEventDto,
  UiDebugFrameDto,
  UiDebugTimelineItemDto,
  UiDebugVariableDto,
  UiDebugVisualElementDto,
  UiSourceSpanDto,
} from "../translation/uiapi";

export const miniCDebugTextFormatterMirror = {
  "javaPath": "src/main/java/minic/uilocal/debug/MiniCDebugTextFormatter.java",
  "webPath": "uiweb/src/debug/MiniCDebugTextFormatter.ts",
  "packageName": "minic.uilocal",
  "exportName": "MiniCDebugTextFormatter",
  "kind": "class",
  "imports": [
    "minic.uiapi.UiDebugBreakpointDto",
    "minic.uiapi.UiDebugEventDto",
    "minic.uiapi.UiDebugFrameDto",
    "minic.uiapi.UiDebugTimelineItemDto",
    "minic.uiapi.UiDebugVariableDto",
    "minic.uiapi.UiDebugVisualElementDto",
    "minic.uiapi.UiSourceSpanDto",
    "java.util.ArrayList",
    "java.util.List",
    "java.util.Map"
  ],
  "fields": [],
  "methods": [
    {
      "name": "addVariableLines",
      "signature": "addVariableLines(List<String> lines, UiDebugVariableDto variable, int depth)"
    },
    {
      "name": "variableTreeText",
      "signature": "variableTreeText(List<UiDebugVariableDto> variables)"
    }
  ]
} as const satisfies JavaMirrorFile;

export class MiniCDebugTextFormatter {
  static readonly mirror = miniCDebugTextFormatterMirror;

  readonly mirror = miniCDebugTextFormatterMirror;

  frameText(frame: UiDebugFrameDto | null | undefined): string {
    if (frame === null || frame === undefined) {
      return this.emptyText("frame");
    }
    return `  ${frame.functionName} return=${frame.returnTarget ?? ""} active=${this.rangeText(frame.activeRange)}`;
  }

  frameWithValuesText(frame: UiDebugFrameDto): string {
    const parameters = this.variableTreeText(frame.parameters);
    const locals = this.variableTreeText(frame.locals);
    return `${this.frameText(frame)} params=[${parameters}] locals=[${locals}]`;
  }

  stackLines(frames: readonly UiDebugFrameDto[]): readonly string[] {
    const lines: string[] = [];
    for (const frame of frames) {
      lines.push(this.frameText(frame));
      if (frame.parameters.length > 0) {
        lines.push("  parameters:");
        lines.push(...this.variableLines(frame.parameters));
      }
      if (frame.locals.length > 0) {
        lines.push("  locals:");
        lines.push(...this.variableLines(frame.locals));
      }
    }
    return lines;
  }

  variableLines(variables: readonly UiDebugVariableDto[]): readonly string[] {
    const lines: string[] = [];
    for (const variable of variables) {
      this.addVariableLines(lines, variable, 0);
    }
    return lines;
  }

  breakpointText(breakpoint: UiDebugBreakpointDto): string {
    return `  line ${breakpoint.line} enabled=${breakpoint.enabled}`;
  }

  eventText(event: UiDebugEventDto): string {
    return `  #${event.eventId} [${event.type}] ${event.title} · ${event.description}`;
  }

  timelineText(item: UiDebugTimelineItemDto): string {
    return `  snapshot ${item.snapshotId} step=${item.visibleStepIndex} reason=${item.stopReason} breakpoint=${item.breakpointHit} range=${this.rangeText(item.sourceRange)}`;
  }

  visualElementText(element: UiDebugVisualElementDto): string {
    const name = element.metadata.fieldName ?? element.label;
    const value = element.metadata.valueSummary ?? "";
    const pointerTarget = element.metadata.pointerTarget ?? "";
    const type = element.metadata.type ?? element.metadata.typeName ?? "";
    if (pointerTarget.trim().length > 0) {
      return `${name}${type.trim().length === 0 ? "" : ` : ${type}`} -> ${pointerTarget}`;
    }
    if (value.trim().length > 0) {
      return `${name}${type.trim().length === 0 ? "" : ` : ${type}`} = ${value}`;
    }
    return `${name}${type.trim().length === 0 ? "" : ` : ${type}`}`;
  }

  emptyText(value: string | null | undefined): string {
    return value === null || value === undefined || value.trim().length === 0 ? "(empty)" : value;
  }

  rangeText(range: UiSourceSpanDto | null | undefined): string {
    if (range === null || range === undefined) {
      return "";
    }
    return `${range.sourceName}:${range.startLine}:${range.startColumn}-${range.endLine}:${range.endColumn}`;
  }

  variableText(variable: UiDebugVariableDto): string {
    return `  ${variable.name} ${variable.typeName} ${variable.valueKind} = ${variable.valueSummary} @ ${variable.address}${
      variable.pointerTarget.trim().length === 0 ? "" : ` pointerTarget=${variable.pointerTarget}`
    }${variable.typeShape.trim().length === 0 ? "" : ` shape=${variable.typeShape}`}${variable.highlightedChange ? " changed" : ""}${
      variable.explanation.trim().length === 0 ? "" : ` · ${variable.explanation}`
    }`;
  }

  addVariableLines(lines: string[], variable: UiDebugVariableDto, depth: number): void {
    lines.push(`  ${"  ".repeat(depth)}${this.variableText(variable).trimStart()}`);
    variable.fields.forEach((field) => this.addVariableLines(lines, field, depth + 1));
    variable.elements.forEach((element) => this.addVariableLines(lines, element, depth + 1));
  }

  variableTreeText(variables: readonly UiDebugVariableDto[]): string {
    return this.variableLines(variables).join("\n");
  }

  summary(): string {
    return `MiniCDebugTextFormatter: ${this.mirror.methods.length} methods, ${this.mirror.fields.length} fields`;
  }
}

export default MiniCDebugTextFormatter;

File: src/main/java/minic/uilocal/debug/MiniCDebugTextFormatter.java
package minic.uilocal;

import minic.uiapi.UiDebugBreakpointDto;
import minic.uiapi.UiDebugEventDto;
import minic.uiapi.UiDebugFrameDto;
import minic.uiapi.UiDebugTimelineItemDto;
import minic.uiapi.UiDebugVariableDto;
import minic.uiapi.UiDebugVisualElementDto;
import minic.uiapi.UiSourceSpanDto;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

final class MiniCDebugTextFormatter {
    private MiniCDebugTextFormatter() {
    }

    static String frameText(UiDebugFrameDto frame) {
        return "  " + frame.functionName()
                + " return=" + (frame.returnTarget() == null ? "" : frame.returnTarget())
                + " active=" + rangeText(frame.activeRange());
    }

    static String frameWithValuesText(UiDebugFrameDto frame) {
        String parameters = variableTreeText(frame.parameters());
        String locals = variableTreeText(frame.locals());
        return frameText(frame)
                + " params=[" + parameters + "]"
                + " locals=[" + locals + "]";
    }

    static List<String> stackLines(List<UiDebugFrameDto> frames) {
        ArrayList<String> lines = new ArrayList<>();
        for (UiDebugFrameDto frame : frames) {
            lines.add(frameText(frame));
            if (!frame.parameters().isEmpty()) {
                lines.add("  parameters:");
                lines.addAll(variableLines(frame.parameters()));
            }
            if (!frame.locals().isEmpty()) {
                lines.add("  locals:");
                lines.addAll(variableLines(frame.locals()));
            }
        }
        return lines;
    }

    static List<String> variableLines(List<UiDebugVariableDto> variables) {
        ArrayList<String> lines = new ArrayList<>();
        variables.forEach(variable -> addVariableLines(lines, variable, 0));
        return lines;
    }

    static String breakpointText(UiDebugBreakpointDto breakpoint) {
        return "  line " + breakpoint.line() + " enabled=" + breakpoint.enabled();
    }

    static String eventText(UiDebugEventDto event) {
        return "  #" + event.eventId()
                + " [" + event.type() + "] " + event.title()
                + " · " + event.description();
    }

    static String timelineText(UiDebugTimelineItemDto item) {
        return "  snapshot " + item.snapshotId()
                + " step=" + item.visibleStepIndex()
                + " reason=" + item.stopReason()
                + " breakpoint=" + item.breakpointHit()
                + " range=" + rangeText(item.sourceRange());
    }

    static String visualElementText(UiDebugVisualElementDto element) {
        Map<String, String> metadata = element.metadata();
        String name = metadata.getOrDefault("fieldName", element.label());
        String value = metadata.getOrDefault("valueSummary", "");
        String pointerTarget = metadata.getOrDefault("pointerTarget", "");
        String type = metadata.getOrDefault("type", metadata.getOrDefault("typeName", ""));
        if (!pointerTarget.isBlank()) {
            return name + (type.isBlank() ? "" : " : " + type) + " -> " + pointerTarget;
        }
        if (!value.isBlank()) {
            return name + (type.isBlank() ? "" : " : " + type) + " = " + value;
        }
        return name + (type.isBlank() ? "" : " : " + type);
    }

    static String emptyText(String value) {
        return value == null || value.isBlank() ? "(empty)" : value;
    }

    static String rangeText(UiSourceSpanDto range) {
        if (range == null) {
            return "";
        }
        return range.sourceName()
                + ":" + range.startLine()
                + ":" + range.startColumn()
                + "-" + range.endLine()
                + ":" + range.endColumn();
    }

    static String variableText(UiDebugVariableDto variable) {
        return "  " + variable.name()
                + " " + variable.typeName()
                + " " + variable.valueKind()
                + " = " + variable.valueSummary()
                + " @ " + variable.address()
                + (variable.pointerTarget().isBlank() ? "" : " pointerTarget=" + variable.pointerTarget())
                + (variable.typeShape().isBlank() ? "" : " shape=" + variable.typeShape())
                + (variable.highlightedChange() ? " changed" : "")
                + (variable.explanation().isBlank() ? "" : " · " + variable.explanation());
    }

    private static void addVariableLines(List<String> lines, UiDebugVariableDto variable, int depth) {
        lines.add("  " + "  ".repeat(depth) + variableText(variable).stripLeading());
        variable.fields().forEach(field -> addVariableLines(lines, field, depth + 1));
        variable.elements().forEach(element -> addVariableLines(lines, element, depth + 1));
    }

    private static String variableTreeText(List<UiDebugVariableDto> variables) {
        return String.join("\n", variableLines(variables));
    }
}
